### ISSUE TITLE
CIRCULAR REFERENCE:com.amazonaws.encryptionsdk.exception.CannotUnwrapDataKeyException: Unable to decrypt any data keys

### ISSUE THREAD

**Issue by anataliocs** [awslabs/aws-encryption-sdk-java#14]
We are seeing the following error when attempting to run a Lambda on AWS.  This exact Lambda was running fine for us, then we wanted to reuse it and updated some of the keys and field/class names then we started getting an encryption key exception.

We have attempted to regenerate the key and the issue persists.  Can you point us in any direction to begin debugging this?

```
Caused by: com.amazonaws.encryptionsdk.exception.CannotUnwrapDataKeyException: Unable to decrypt any data keys
at com.amazonaws.encryptionsdk.MasterKeyProvider.buildCannotDecryptDksException(MasterKeyProvider.java:106)
at com.amazonaws.encryptionsdk.multi.MultipleProviderFactory$MultiProvider.decryptDataKey(MultipleProviderFactory.java:158)
at com.amazonaws.encryptionsdk.internal.DecryptionHandler.getDataKey(DecryptionHandler.java:329)
at com.amazonaws.encryptionsdk.internal.DecryptionHandler.readHeaderFields(DecryptionHandler.java:382)
at com.amazonaws.encryptionsdk.internal.DecryptionHandler.<init>(DecryptionHandler.java:124)
at com.amazonaws.encryptionsdk.AwsCrypto.decryptData(AwsCrypto.java:242)
at com.amazonaws.encryptionsdk.AwsCrypto.decryptData(AwsCrypto.java:233)
at com.nike.webhooks.common.security.KmsEncryptDecrypt.decrypt(KmsEncryptDecrypt.java:37)
at com.nike.webhooks.common.repository.EventConverter.unconvert(EventConverter.java:55)
at com.nike.webhooks.common.repository.EventConverter.unconvert(EventConverter.java:18)
at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBTypeConverter$DelegateConverter.unconvert(DynamoDBTypeConverter.java:109)
at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBTypeConverter$NullSafeConverter.unconvert(DynamoDBTypeConverter.java:128)
at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBTypeConverter$ExtendedConverter.unconvert(DynamoDBTypeConverter.java:88)
at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapperFieldModel.unconvert(DynamoDBMapperFieldModel.java:146)
at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapperFieldModel.unconvertAndSet(DynamoDBMapperFieldModel.java:164)
at com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapperTableModel.unconvert(DynamoDBMapperTableModel.java:267)
... 18 more
Suppressed: com.amazonaws.encryptionsdk.exception.CannotUnwrapDataKeyException: Unable to decrypt any data keys
at com.amazonaws.encryptionsdk.MasterKeyProvider.buildCannotDecryptDksException(MasterKeyProvider.java:103)
at com.amazonaws.encryptionsdk.kms.KmsMasterKey.decryptDataKey(KmsMasterKey.java:182)
at com.amazonaws.encryptionsdk.multi.MultipleProviderFactory$MultiProvider.decryptDataKey(MultipleProviderFactory.java:150)
... 32 more
[CIRCULAR REFERENCE:com.amazonaws.encryptionsdk.exception.CannotUnwrapDataKeyException: Unable to decrypt any data keys]
```

**Comment by SalusaSecondus** [comment]
Double check how you're using the MultipleProviderFactory. I'm guessing you've accidentally created a loop.

**Comment by anataliocs** [comment]
@SalusaSecondus Thanks for getting back to us so quickly.

This is how we are using MultipleProviderFactory.  We pass in a String array of keys, get an instance of KmsMasterKey for each one then pass in those keys to MultipleProviderFactory.buildMultiProvider().

Do you see any potential issues with that? 

```
public class KmsEncryptDecrypt implements EncryptDecrypt {

    public byte[] encrypt(byte[] bytes, String[] keyArns) {
        MasterKeyProvider<?> masterKeyProvider = this.initializeMasterKeyProvider(keyArns);
        AwsCrypto crypto = new AwsCrypto();
        return crypto.encryptData(masterKeyProvider, bytes).getResult();
    }

    @Override
    public byte[] encrypt(byte[] bytes, String keyArn) {
        return this.encrypt(bytes, new String[]{keyArn});
    }

    @Override
    public byte[] decrypt(byte[] bytes, String keyArn) {
        return this.decrypt(bytes, new String[]{keyArn});
    }

    public byte[] decrypt(byte[] cipherBytes, String[] keyArns) {
        MasterKeyProvider<?> masterKeyProvider = this.initializeMasterKeyProvider(keyArns);
        AwsCrypto crypto = new AwsCrypto();
        return crypto.decryptData(masterKeyProvider, cipherBytes).getResult();
    }

    private MasterKeyProvider<?> initializeMasterKeyProvider(String[] kmsKeyArns) {
        DefaultAWSCredentialsProviderChain chain = new DefaultAWSCredentialsProviderChain();

        List<KmsMasterKey> keys = Stream.of(kmsKeyArns)
                .filter(s -> s.trim().length() > 0)
                .map(s -> KmsMasterKey.getInstance(chain, s))
                .collect(Collectors.toList());

        if(keys.size() == 0) {
            throw new IllegalArgumentException("No kms key arns provided");
        }
        return MultipleProviderFactory.buildMultiProvider(keys);
    }
}
```

**Comment by SalusaSecondus** [comment]
I'll dig into this a bit more. FYI, the decrypt logic might be a bit off from a security standpoint. You should check the CryptoResult to ensure that the key used to decrypt is one of the ones you passed in to the decrypt method. (Depending on how you've set up your AWS credentials and permissions there are some edge-cases where this could be a problem.)

**Comment by SalusaSecondus** [comment]
Okay. I see the bug in the code and will get it fixed. This won't fix your underlying problems with decryption, but will fix the stack-trace.

**Comment by pdhruv** [comment]
The KMS key is setup to allow 2 applications to use it. 

The first application uses the encrypt the content using the KMS key ARN, performs Base64 encoding and persists it in dynamodb.

The second application retrieves data stored in dynamodb, performs Base64 decode and then decrypts the content using the same KMS key ARN.

Do you see any issue with how we are using the key?

**Comment by SalusaSecondus** [comment]
Unfortunately, without the corrected stacktrace, I don't know what's wrong. You should double check permissions on both the key and the principals used by your applications when they call KMS.

**Comment by anataliocs** [comment]
@SalusaSecondus It looks like the error is sporadic and occurs based on how the key is created.  Using an older existing key with our Lambda, we are able to get it working.  Using newly generated keys is when we see this issue.

Could this be related to:  https://github.com/awslabs/aws-encryption-sdk-java/issues/10

**Comment by SalusaSecondus** [comment]
No, this isn't related to #10 which has to do with ciphertext formats. However, generating the keys differently may have resulted in different key policies. Check those out as they'll define the permissions associated with the keys.

**Comment by SalusaSecondus** [comment]
Fixed: f108896d8c34c7f0c94590ea2a80971133ceb1b6
